File: 2024/lib/day01.erl
-module(day01).

-export([ solve/1 ]).

split([]) -> {[], []};
split([A, B | Rest]) ->
    {As, Bs} = split(Rest),
    {[A | As], [B | Bs]}.

solve(Input) ->
    {A, B} = split(utils:ints(Input)),
    Zipped = lists:zip(lists:sort(A), lists:sort(B), trim),
    { lists:sum([abs(I - J) || {I, J} <- Zipped])
    , lists:sum([I * length(lists:filter(fun(J) -> J == I end, B)) || I <- A])
    }.

File: 2021/lib/Day01.ex
defmodule Day01 do
  def solve(input) do
    parsed = input |> H.ints()

    ans1 =
      parsed
      |> Enum.zip(tl(parsed))
      |> Enum.filter(fn {a, b} -> b > a end)
      |> Enum.count()

    ans2 =
      Enum.zip([parsed, tl(parsed), tl(tl(parsed))])
      |> Enum.map(fn {a, b, c} -> a + b + c end)
      |> then(fn l -> Enum.zip(l, tl(l)) end)
      |> Enum.filter(fn {a, b} -> b > a end)
      |> Enum.count()

    {"#{ans1}", "#{ans2}"}
  end
end
